src/storage/src/hummock/local_version_manager.rs

```diff
@@ -554,27 +616,132 @@ impl LocalVersionManager {
+             epoch_join_handle: &mut HashMap<HummockEpoch, Vec<JoinHandle<()>>>| {
+                // Keep issuing new flush task until flush is not needed or we can issue
+                // no more task
+                while local_version_manager.buffer_tracker.need_more_flush() {
+                    if let Some((epoch, join_handle)) = local_version_manager
+                        .clone()
+                        .flush_shared_buffer(syncing_epoch)
+                    {
+                        epoch_join_handle
+                            .entry(epoch)
+                            .or_insert_with(Vec::new)
+                            .push(join_handle);
+                    } else {
+                        break;
+                    }
+                }
+            };
+
+        let grant_write_request = |request| {
+            let WriteRequest {
+                batch,
+                epoch,
+                is_remote_batch,
+                grant_sender: sender,
+            } = request;
+            let size = batch.size();
+            local_version_manager.write_shared_buffer_inner(epoch, batch, is_remote_batch);
+            local_version_manager
+                .buffer_tracker
+                .global_buffer_size
+                .fetch_add(size, Relaxed);
+            let _ = sender.send(()).inspect_err(|err| {
+                error!("unable to send write request response: {:?}", err);
+            });
+        };
+
+        let mut pending_write_requests: VecDeque<_> = VecDeque::new();
+
+        // While the current Arc is not the only strong reference to the local version manager
+        while Arc::strong_count(&local_version_manager) > 1 {
+            if let Some(event) = buffer_size_change_receiver.recv().await {
+                match event {
+                    SharedBufferEvent::WriteRequest(request) => {
+                        if local_version_manager.buffer_tracker.can_write() {
+                            grant_write_request(request);
+                            try_flush_shared_buffer(&syncing_epoch, &mut epoch_join_handle);
+                        } else {
+                            pending_write_requests.push_back(request);
+                        }
+                    }
+                    SharedBufferEvent::MayFlush => {
+                        // Only check and flush shared buffer after batch has been added to shared
+                        // buffer.
+                        try_flush_shared_buffer(&syncing_epoch, &mut epoch_join_handle);
+                    }
+                    SharedBufferEvent::BufferRelease(size) => {
+                        local_version_manager
+                            .buffer_tracker
+                            .global_buffer_size
+                            .fetch_sub(size, Relaxed);
+                        while !pending_write_requests.is_empty()
+                            && local_version_manager.buffer_tracker.can_write()
```

Comment:
Why we must check `can_write` before trigger a new flush task? If the memory exceeds the block threshold, we can not flush data, can we?

Reply:
`can_write` is checked to decide whether to grant the pending write request stored in `pending_write_requests`. When a write request is granted and new data is written to shared buffer, the `global_buffer_size` may exceed the threshold again and we may trigger the flush.

When memory exceeds block threshold, we need flush data to free some memory space to unblock the write.